[libpas] Use __thread keyword on Linux and FreeBSD

Constellation opened the pull request:
#### 3d8b0c6d5f0aae9a197be0b2075b7a2625ac3d7d
<pre>
[libpas] Use __thread keyword on Linux and FreeBSD
<a href="https://bugs.webkit.org/show_bug.cgi?id=243975">https://bugs.webkit.org/show_bug.cgi?id=243975</a>

Reviewed by Mark Lam.

This patch adds a new handling of TLS in libpas: using __thread for Linux and FreeBSD.
The mechanism is that,

1. Declare __thread with initial value NULL.
2. Then, when setting a valid TLS value, we also call pthread_setspecific to schedule thread exit callback.
3. In thread exit callback, we clear __thread with PAS_THREAD_LOCAL_CACHE_DESTROYED.
4. So, we can check whether the thread is dying by looking at __thread&apos;s variable, if it is PAS_THREAD_LOCAL_CACHE_DESTROYED, then it is dying.

This offers better performance in Linux and FreeBSD since we can use ELF&apos;s TLS allocation for libpas TLC.
On OSS Darwin, we continue using repeated pthread_setspecific since __thread implementation does not work. This is because it can be cleared to the initial
value after the TLS clearing iteration. So we cannot use this value for pthread exiting detection.

* Source/bmalloc/libpas/src/libpas/pas_fast_tls.h:
* Source/bmalloc/libpas/src/libpas/pas_thread_local_cache.c:
(destructor):
* Source/bmalloc/libpas/src/libpas/pas_thread_local_cache.h:
(pas_thread_local_cache_try_get):
(pas_thread_local_cache_can_set):
* Source/bmalloc/libpas/src/libpas/pas_utils.h:

Canonical link: <a href="https://commits.webkit.org/253564@main">https://commits.webkit.org/253564@main</a>
</pre>









<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5c5223f471aeaec45a8f833ac92ffcc24a747b20

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86365 "failed 3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30285 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17334 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95209 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/148922 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/28655 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/25297 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/78501 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90461 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/91965 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23242 "Passed tests") | [✅ ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/73359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23338 "Passed tests") | 
| | [✅ ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/78252 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/78676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66343 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/78309 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/26604 "Built successfully") | [✅ ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/12525 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/71936 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/26514 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/13539 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25726 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2539 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28191 "Built successfully") | [✅ ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/73359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/74717 "failed Failed to checkout and rebase branch from PR 3345") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28131 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/32820 "Passed tests") | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/37/builds/74717 "failed Failed to checkout and rebase branch from PR 3345") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_thread_local_cache.h:
```diff
@@ -118,9 +126,13 @@ static inline bool pas_thread_local_cache_can_set(void)
 #ifdef PAS_THREAD_LOCAL_CACHE_CAN_DETECT_THREAD_EXIT
     return !pthread_self_is_exiting_np();
+#else
+#if PAS_HAVE_THREAD_KEYWORD
```
Why not use `#elif PAS_HAVE_THREAD_KEYWORD` instead?

Constellation replied:
Fixed.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_thread_local_cache.h:
```diff
@@ -118,9 +126,13 @@ static inline bool pas_thread_local_cache_can_set(void)
 #else
     return ((uintptr_t)pas_thread_local_cache_try_get_impl()) != PAS_THREAD_LOCAL_CACHE_DESTROYED;
 #endif
+#endif
```
Not needed if you use `#elif` above.

Constellation replied:
Dropped.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_fast_tls.h:
```diff
@@ -84,20 +84,17 @@ struct pas_fast_tls {
-    })
-
-#define PAS_FAST_TLS_SET(static_key, passed_fast_tls, value) do { \
+/* Regardless of whether the state of fast_tls, we can access static_key. */
```
I suggesting adding a bit more detail to the comment e.g.
```
/* This is the PAS_HAVE_THREAD_KEYWORD case. Hence, static_key here is not actually a key, but a thread
   local cache of the value (declared with the __thread attribute). Regardless of whether fast_tls has been
   initialized yet or not, it is safe to access this thread local cache of the value. */
```
Am I understanding this correctly?

Constellation replied:
Yes.

MenloDorian reviewed (approved):
r=me with suggestions.

basuke commented on Source/bmalloc/libpas/src/libpas/pas_fast_tls.h:
```diff
@@ -84,20 +84,17 @@ struct pas_fast_tls {
         } \
     } while (false)
 
-#define PAS_FAST_TLS_GET(static_key, passed_fast_tls) ({ \
```
Should we keep the original implementation for the case when __thread is not exist? I'm not sure we should care about that situation, but

Constellation replied:
Yes, we will keep it since it turned out that this is necessary for Darwin.

basuke commented on Source/bmalloc/libpas/src/libpas/pas_thread_local_cache.h:
```diff
@@ -118,9 +126,13 @@ static inline bool pas_thread_local_cache_can_set(void)
+#if PAS_HAVE_THREAD_KEYWORD
+    return ((uintptr_t)pas_thread_local_cache_try_get_impl()) >= PAS_THREAD_LOCAL_CACHE_UNINITIALIZED;
 #else
     return ((uintptr_t)pas_thread_local_cache_try_get_impl()) != PAS_THREAD_LOCAL_CACHE_DESTROYED;
```
...still there's an implementation for non PAS_HAVE_THREAD_KEYWORD case.

Constellation replied:
Older internal macOS does not have `PAS_THREAD_LOCAL_CACHE_CAN_DETECT_THREAD_EXIT`. So we need to keep it regardless of whether `PAS_HAVE_THREAD_KEYWORD` exists.

basuke reviewed (commented):
So there're many conditions mixed:
- PAS_HAVE_PTHREAD_MACHDEP_H
- PAS_THREAD_LOCAL_CACHE_CAN_DETECT_THREAD_EXIT
- PAS_HAVE_THREAD_KEYWORD

Other than apple port, first two are off. The third one is currently defined as !PAS_HAVE_PTHREAD_MACHDEP_H, but it is possible to define that more precisely manually.

So I think we have two choices.
- Keep original implementation of `PAS_FAST_TLS_GET|SET` for the backup of PAS_HAVE_THREAD_KEYWORD = off case.
- Remove PAS_HAVE_THREAD_KEYWORD and make it requirement for non-apple build.

basuke commented on Source/bmalloc/libpas/src/libpas/pas_fast_tls.h:
```diff
@@ -84,20 +84,17 @@ struct pas_fast_tls {
-        pthread_setspecific(fast_tls->key, (value)); \
+        static_key = value; \
+        if (((uintptr_t)value) != PAS_THREAD_LOCAL_CACHE_DESTROYED) \
+            pthread_setspecific(fast_tls->key, value); \
```
This is required for calling destructor, right? Should we leave a comment for that fact?

Constellation replied:
Added

Constellation committed:
[libpas] Use __thread keyword on Linux and FreeBSD
https://bugs.webkit.org/show_bug.cgi?id=243975

Reviewed by Mark Lam.

This patch adds a new handling of TLS in libpas: using __thread for Linux and FreeBSD.
The mechanism is that,

1. Declare __thread with initial value NULL.
2. Then, when setting a valid TLS value, we also call pthread_setspecific to schedule thread exit callback.
3. In thread exit callback, we clear __thread with PAS_THREAD_LOCAL_CACHE_DESTROYED.
4. So, we can check whether the thread is dying by looking at __thread's variable, if it is PAS_THREAD_LOCAL_CACHE_DESTROYED, then it is dying.

This offers better performance in Linux and FreeBSD since we can use ELF's TLS allocation for libpas TLC.
On OSS Darwin, we continue using repeated pthread_setspecific since __thread implementation does not work. This is because it can be cleared to the initial
value after the TLS clearing iteration. So we cannot use this value for pthread exiting detection.

* Source/bmalloc/libpas/src/libpas/pas_fast_tls.h:
* Source/bmalloc/libpas/src/libpas/pas_thread_local_cache.c:
(destructor):
* Source/bmalloc/libpas/src/libpas/pas_thread_local_cache.h:
(pas_thread_local_cache_try_get):
(pas_thread_local_cache_can_set):
* Source/bmalloc/libpas/src/libpas/pas_utils.h:

Canonical link: https://commits.webkit.org/253564@main
